compactor2/src/components/split_or_compact/start_level_files_to_split.rs

```diff
@@ -7,12 +7,161 @@ use crate::{
+///    - "L0.60[42,986] 1.05us 10mb   |---------------------------------------L0.60---------------------------------------|  "
+///
+///  . Output:
+///     split all L0 input files at: split(split_times=[248, 372, 496, 620, 744, 868]
+///
+/// Reason behind the split:
+///  If this input was allowed to compact and split based on the algorithm focused on less
+///  extensively overlapped files, it can produce overlapped L1s so large than they can
+///  only be compacted 2-3 at a time, and then resplit at their prior boundaries.
+///
+///  Instead, this algorithm recognizes the large quantity of highly overlapped data and
+///  performs a 'vertical split', instead of selecting bite sized set of files and splitting
+///  just them, it selects all overlapping data and splits all of the files at the same
+///  time(s).  A single split decision made by this function may split a quantity of files
+///  many times larger than what we can compact in a single compaction.  But the split will
+///  allow a future compaction to fit all the data for a relatively narrow time range.
+///
+pub fn high_l0_overlap_split(
+    max_compact_size: usize,
+    files: Vec<ParquetFile>,
+    target_level: CompactionLevel,
+) -> (Vec<FileToSplit>, Vec<ParquetFile>) {
+    let start_level = target_level.prev();
+    let len = files.len();
+
+    // This function focuses on many overlaps in a single level, which is only possible in L0.
+    if start_level != CompactionLevel::Initial {
+        return (vec![], files);
+    }
+
+    // We'll apply vertical splitting when the number of overlapped files is too large
+    // for a single compaction.
+    let total_cap: usize = files.iter().map(|f| f.file_size_bytes as usize).sum();
+
+    if total_cap < max_compact_size {
+        // Take no action, pass all files on to the next rule.
+        return (vec![], files);
+    }
+
+    // panic if not all files are either in target level or start level
+    assert!(files
+        .iter()
+        .all(|f| f.compaction_level == target_level || f.compaction_level == start_level));
+
+    // Get files in start level that overlap with any file in target level
+    let mut files_to_split = Vec::with_capacity(len);
+    let mut overlaps: Vec<&ParquetFile> = Vec::with_capacity(len);
+    let mut files_not_to_split = Vec::with_capacity(len);
+
+    // The files in `files` all overlap, but they might overlap in a chain (file1 overlaps file1, which overlaps file3...)
+    // This algorithm is not concerned with the chain of overlaps that keeps pulling in more files.  Instead, for each file,
+    // we'll look at what overlaps just it.  If that's over max_compact_size, we'll split them.
+    for file in &files {
+        if file.compaction_level != start_level {
+            continue;
+        }
+
+        // Get the set of overlaping start_level files that includes file
+        overlaps = files.iter().filter(|f| file.overlaps(f)).collect();
```

Comment:
Since `files` also include the `file`, I think this overlaps will include all files, right? Which means not what you want? Maybe you have to check `f != file`, too
```suggestion
        overlaps = files.iter().filter(|f|  f!file &&   file.overlaps(f)).collect();
```

I have not verified if it is compiled

Reply:
If so so, `overlaps` is actually the full set of `files` with start_level, right?
If that is what you need, you only need iterator with a filter to keep only start_level files

Reply:
> If so so, `overlaps` is actually the full set of `files` with start_level, right?

no.
`files` is the full chain of overlapping files.

For example, for this:
```
        - "L0.2[650,750] 180s                                                    |------L0.2------|                           "
        - "L0.1[450,620] 120s                |------------L0.1------------|                                                   "
        - "L0.3[800,900] 300s                                                                               |------L0.3------|"
        - "L1, all files 100b                                                                                                 "
        - "L1.13[600,700] 60s                                           |-----L1.13------|                                    "
        - "L1.12[400,500] 60s       |-----L1.12------|                                                                        "
```
`files` could be L1.12, L0.1, L1.13, L0.2 I call that a "chain" because L1.12 overlaps L0.1, which overlaps L1,13, which overlaps L0.2.   L1.12 does not overlap L0.2, but they're in `files` together because a chain of overlaps that doesn't break until L0.3, which doesn't overlap anything.

This algorithm selects narrower sets of files, not the whole chain.
So if the for loop starts with L0.2, `overlaps` would contain only L0.2 and L1.13 because L1.13 is the only thing that overlaps L0.2.  

Think of the logic as:
If following the chain pulls in too much and we can't make good decisions on it, instead of following the chain we should pick an L0 and focus just on what we need to for it.

Reply:
The explanation makes sense. Thanks @Joe-Blount 

I have  different question/comment. Let us look at the example above and which is also the test I commented above. Why in that test, L0.1 (and its overlapped L1.12) is not selected but instead L0.2 (and its overlapped L1.13) is selected. I guess because L0.2 happens to be in front of L0.1 in the `files`. I think since we pick files with smallest `max_l0_created_at` to compact first, splitting those files first will provide higher chance to reduce the number of rounds to compact and split. 

What do you think to order `files` in the order of `max_l0_created_at` first and then run the loop to select what to split?

Reply:
Great point! yes, I should sort the files before the loop.